dashboard/modules/log/log_head.py

```diff
@@ -100,3 +106,193 @@ async def run(self, server):
+        if change.old:
+            node_id, _ = change.old
+            ip = DataSource.node_id_to_ip[node_id]
+            self._stubs.pop(node_id)
+            self._ip_to_node_id.pop(ip)
+        if change.new:
+            node_id, ports = change.new
+            ip = DataSource.node_id_to_ip[node_id]
+
+            options = (("grpc.enable_http_proxy", 0),)
+            channel = init_grpc_channel(
+                f"{ip}:{ports[1]}", options=options, asynchronous=True
+            )
+            stub = reporter_pb2_grpc.LogServiceStub(channel)
+            self._stubs[node_id] = stub
+            self._ip_to_node_id[ip] = node_id
+
+    @staticmethod
+    async def get_logs_json_index(grpc_stub, filters):
+        reply = await grpc_stub.LogIndex(reporter_pb2.LogIndexRequest())
+        filters = [] if filters == [""] else filters
+        links = list(filter(lambda s: all(f in s for f in filters), reply.log_files))
+        logs = {}
+        logs["worker_errors"] = list(
+            filter(lambda s: "worker" in s and s.endswith(".err"), links)
+        )
+        logs["worker_outs"] = list(
+            filter(lambda s: "worker" in s and s.endswith(".out"), links)
+        )
+        for lang in ray_constants.LANGUAGE_WORKER_TYPES:
+            logs[f"{lang}_core_worker_logs"] = list(
+                filter(
+                    lambda s: f"{lang}-core-worker" in s and s.endswith(".log"),
+                    links,
+                )
+            )
+            logs[f"{lang}_driver_logs"] = list(
+                filter(
+                    lambda s: f"{lang}-core-driver" in s and s.endswith(".log"),
+                    links,
+                )
+            )
+        logs["dashboard"] = list(filter(lambda s: "dashboard" in s, links))
+        logs["raylet_logs"] = list(filter(lambda s: "raylet" in s, links))
+        logs["gcs_logs"] = list(filter(lambda s: "gcs" in s, links))
+        logs["ray_client"] = list(filter(lambda s: "ray_client" in s, links))
+        logs["autoscaler_monitor"] = list(
+            filter(lambda s: "monitor" in s and "log_monitor" not in s, links)
+        )
+        logs["folders"] = list(filter(lambda s: "." not in s, links))
+        logs["misc"] = list(
+            filter(lambda s: all([s not in logs[k] for k in logs]), links)
+        )
+        return logs
+
+    @routes.get("/api/experimental/logs/index")
+    async def handle_log_index(self, req):
+        node_id_query = req.query.get("node_id", None)
+        filters = req.query.get("filters", "").split(",")
+        response = await self.get_log_index(node_id_query, filters)
+        return aiohttp.web.json_response(response)
+
+    async def get_log_index(self, node_id_query, filters):
+        response = {}
+        while self._stubs == {}:
+            await asyncio.sleep(0.5)
+        tasks = []
+        for node_id, grpc_stub in self._stubs.items():
+            if node_id_query is None or node_id_query == node_id:
```

Comment:
I think if no node id is given, the user should expect the logs index of the whole cluster.

Reply:
I think it is useful for SDK and dashboard consumers.

For CLI, I paginate to 10 nodes.

Reply:
In this case, can you always put the head node node id first and add a test?

Reply:
Wait, I don't understand. So isn't the conclusion that we should display the full index if no node id is specified? Since that is probably what SDK and dashboard want. 

If we want to get head node's, it should be easy to get via its node id.

Reply:
Ok, so we should have head node's list be sorted as the first entry in the json...

Reply:
It's quite hard to reorder json manually. Also I think we are not guaranteed that consumer will get the dict in order when they do json.loads. Hence, I think we should leave it.